src/core/thread/mlr_manager.cpp

```diff
@@ -0,0 +1,359 @@
+    VerifyOrExit(aMessage->GetCode() == OT_COAP_CODE_CHANGED, error = OT_ERROR_PARSE);
+
+    SuccessOrExit(error = Tlv::FindUint8Tlv(*aMessage, ThreadTlv::kStatus, status));
+
+exit:
+
+    if (status == ThreadStatusTlv::MlrStatus::kMlrSuccess)
+    {
+        mTable.UpdateRegisteringSubscriptions(true);
+        // keep sending until all multicast addresses are registered.
+        ScheduleSend(0);
+    }
+    else
+    {
+        // Conservatively schedule reregistration if failed.
+        // This is required by Test Spec. MATN-TC-26
+        // TODO (MLR): reregister only failed addresses.
+        mTable.UpdateRegisteringSubscriptions(false);
+        UpdateReregistrationPeriod(true);
+    }
+
+    otLogInfoBbr("Receive MLR.rsp, result=%s, status=%d, error=%s", otThreadErrorToString(aResult), status,
+                 otThreadErrorToString(error));
+}
+
+void MlrManager::HandleTimer(void)
+{
+    if (mSendDelay > 0)
+    {
+        if (--mSendDelay == 0)
+        {
+            SendMulticastListenerRegistration();
+        }
+    }
+
+    if (mReregistrationDelay > 0)
+    {
+        if (--mReregistrationDelay == 0)
+        {
+            Reregistration();
+        }
+    }
+
+    ResetTimer();
+}
+
+void MlrManager::LogMulticastSubscriptionTable(void)
+{
+    mTable.Print();
+}
+
+void MlrManager::Reregistration(void)
+{
+    mTable.ClearAllRegistered();
+    ScheduleSend(0);
+
+    // Schedule for the next renewing.
+    UpdateReregistrationPeriod(false);
+}
+
+void MlrManager::UpdateReregistrationPeriod(bool aRereg)
+{
+    Mle::MleRouter &mle = Get<Mle::MleRouter>();
+
+    bool needSendMlr = (mle.IsFullThreadDevice() || mle.GetParent().IsThreadVersion1p1()) &&
+                       Get<BackboneRouter::Leader>().HasPrimary();
+
+    if (!needSendMlr)
+    {
+        mReregistrationDelay = 0;
+    }
+    else
+    {
+        otBackboneRouterConfig config;
+        uint32_t               reregDelay;
+        uint32_t               effectiveMlrTimeout;
+        uint32_t               minPeriod, maxPeriod;
+
+        IgnoreError(Get<BackboneRouter::Leader>().GetConfig(config));
+
+        if (aRereg)
+        {
+            reregDelay = config.mReregistrationDelay > 1
+                             ? Random::NonCrypto::GetUint16InRange(1, config.mReregistrationDelay)
+                             : 1;
+        }
+        else
+        {
+            // Calculate renewing period according to Thread Spec. 5.24.2.3.2
+            // The random time t SHOULD be chosen such that (0.5* MLR-Timeout) < t < (MLR-Timeout – 9 seconds).
+            effectiveMlrTimeout = config.mMlrTimeout > Mle::kMlrTimeoutMin ? config.mMlrTimeout
+                                                                           : static_cast<uint32_t>(Mle::kMlrTimeoutMin);
+            minPeriod  = (effectiveMlrTimeout >> 1u) + 1;
+            maxPeriod  = effectiveMlrTimeout - 9 > minPeriod ? effectiveMlrTimeout - 9 : minPeriod + 1;
```

Comment:
Note that `GetUint32InRange` will crash if `maxPeriod` equals to `minPeriod`. 

`minPeriod  = (effectiveMlrTimeout >> 1u) + 1;` is used because the random t must satisfy `t > (0.5* MLR-Timeout)`.

The variable name `maxPeriod` is a little misleading because `GetUint32InRange` returns a number `minPeriod <= random value < maxPeriod`. 
I think the actual max period should be `max(minPeriod, effectiveMlrTimeout - 10)`.
So, the current code is indeed **correct**.

Reply:
I think I am over thinking because since `effectiveMlrTimeout >= 300`, `effectiveMlrTimeout - 9 > minPeriod` should always be satisfied. I will simplify the code.